In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from db import engine

plt.rcParams["figure.figsize"] = (15, 6)
sns.set_style('whitegrid')

In [2]:
with engine.connect() as conn:
    loc_df = pd.read_sql("select * from location", conn)
    weather_df = pd.read_sql("select * from weather_observation", conn)

In [3]:
df = loc_df.merge(weather_df, left_on='id', right_on='location_id', suffixes=['_loc', '_wea'])

In [4]:
df.sample(10)

,id_loc,city,state,country,latitude,longitude,created_at_loc,id_wea,location_id,observed_at,temperature,feels_like,humidity,pressure,wind_speed,weather_description,weather_code,raw_json,created_at_wea
341495,14840ebd-d24e-4d4c-bc5f-9826dc2b8ae0,Las Vegas,NV,USA,36.1699,-115.1398,2026-05-27 15:58:37.949941+00:00,b6375f15-cc5c-4298-875c-700964f28e95,14840ebd-d24e-4d4c-bc5f-9826dc2b8ae0,2025-12-13 00:00:00+00:00,20.7,17.6,17.0,943.6,2.5,Clear,0,None,2026-05-27 16:15:25.648081+00:00
714711,b2a5c003-35c3-46db-93e0-a0bef17c8d0e,Philadelphia,PA,USA,39.9526,-75.1652,2026-05-27 15:58:37.949941+00:00,cb9bd8d8-cb24-4fa1-b0ef-081cfa6695a0,b2a5c003-35c3-46db-93e0-a0bef17c8d0e,2023-07-07 22:00:00+00:00,27.0,32.0,81.0,1005.9,5.4,Partly cloudy,3,None,2026-05-27 16:15:29.871499+00:00
526592,48d5cbfc-39e1-421a-86c8-01d6a543e645,Dallas,TX,USA,32.7767,-96.7970,2026-05-27 15:58:37.949941+00:00,4b3e9a87-3b3c-4e99-9aea-9e54aebee6bd,48d5cbfc-39e1-421a-86c8-01d6a543e645,2023-01-22 08:00:00+00:00,6.4,2.4,66.0,997.3,13.0,Clear,0,None,2026-05-27 16:15:22.483638+00:00
30869,a7e5ccad-ac93-4b16-8997-23c4daa14103,Bothell,WA,USA,47.7623,-122.2054,2026-05-27 15:58:37.949941+00:00,443a3187-1846-462d-8f37-71f81076b42d,a7e5ccad-ac93-4b16-8997-23c4daa14103,2023-07-10 05:00:00+00:00,17.0,15.7,68.0,1010.4,11.3,Clear,0,None,2026-05-27 16:15:21.576695+00:00
483960,9dea3f12-e331-4152-a029-ef9aa1875dc8,Minneapolis,MN,USA,44.9778,-93.2650,2026-05-27 15:58:37.949941+00:00,fc1bbdbd-13fe-4d3a-8be8-148166e31701,9dea3f12-e331-4152-a029-ef9aa1875dc8,2024-03-16 03:00:00+00:00,7.0,2.5,44.0,981.4,12.3,Partly cloudy,1,None,2026-05-27 16:15:27.230059+00:00
276233,bbd00773-929e-4bdb-b0b9-0f32802d819a,San Diego,CA,USA,32.7157,-117.1611,2026-05-27 15:58:37.949941+00:00,d4edafba-0398-4bb1-9d4a-22af97000d3f,bbd00773-929e-4bdb-b0b9-0f32802d819a,2024-07-03 17:00:00+00:00,21.4,24.4,87.0,1009.7,3.4,Foggy,45,None,2026-05-27 16:15:32.912488+00:00
502000,b58ee725-c52e-4e49-a5f8-db44b6d89f9e,Kansas City,MO,USA,39.0997,-94.5786,2026-05-27 15:58:37.949941+00:00,19c8841c-7b08-4942-8952-17d80a3993a0,b58ee725-c52e-4e49-a5f8-db44b6d89f9e,2023-04-03 16:00:00+00:00,13.2,11.4,76.0,969.5,10.8,Partly cloudy,3,None,2026-05-27 16:15:24.670974+00:00
335637,14840ebd-d24e-4d4c-bc5f-9826dc2b8ae0,Las Vegas,NV,USA,36.1699,-115.1398,2026-05-27 15:58:37.949941+00:00,cabc9a8d-892b-4ab6-a3ee-cb931eac358d,14840ebd-d24e-4d4c-bc5f-9826dc2b8ae0,2025-04-12 22:00:00+00:00,31.8,25.5,7.0,938.6,30.0,Partly cloudy,3,None,2026-05-27 16:15:25.648081+00:00
308503,0b6b84ef-6a18-48ba-9397-46ae64193be9,Phoenix,AZ,USA,33.4484,-112.0740,2026-05-27 15:58:37.949941+00:00,80990109-c8d0-4ade-9d98-5432f97e14ed,0b6b84ef-6a18-48ba-9397-46ae64193be9,2025-03-09 07:00:00+00:00,12.3,9.9,54.0,984.2,5.9,Clear,0,None,2026-05-27 16:15:31.014004+00:00
593786,ae8aab9d-0793-4e34-a7a1-b3fb8e505392,New Orleans,LA,USA,29.9511,-90.0715,2026-05-27 15:58:37.949941+00:00,a6ef1161-e6bf-44ae-b2ec-124a2234d5b0,ae8aab9d-0793-4e34-a7a1-b3fb8e505392,2024-11-18 18:00:00+00:00,27.5,27.8,60.0,1008.9,21.3,Partly cloudy,3,None,2026-05-27 16:15:27.819536+00:00


In [4]:
df.isna().sum()

id_loc                      0
city                        0
state                       0
country                     0
latitude                    0
longitude                   0
created_at_loc              0
id_wea                      0
location_id                 0
observed_at                 0
temperature                 0
feels_like                  0
humidity                    0
pressure                    0
wind_speed                  0
weather_description         0
weather_code                0
raw_json               789120
created_at_wea              0
dtype: int64

In [5]:
df['weather_description'].value_counts(dropna=False)

weather_description
Partly cloudy       366469
Clear               358148
Drizzle              22281
Foggy                16296
Thunderstorm         12087
Rain                  7387
Snow                  5611
Rain Showers           698
Freezing Drizzle       106
Freezing Rain           34
Snow Showers             3
Name: count, dtype: int64

> From the above outputs, we see that we have 30 U.S. cities, with hourly weather data ranging from January 1st, 2023 to December 31st, 2025. We have 3 years worth of data, with each city having the same amount of hourly data.


> The only columns in which there is missing data is the raw_json, which we will drop as this is missing for every row, and weather_description. The missing data in weather description means that the data contained a weather code that was not covered in the data dictionary during the transformation process.
>
> NOTE: originally, there was 12087 missing data values. this was fixed in the database directly

In [6]:
df[df['weather_description'].isna()].value_counts('weather_code')

Series([], Name: count, dtype: int64)

Looking at the OpenMeteo documentation, the weather code "95" corresponds to thunderstorms, so we can fill in the missing values for weather descriptions. 

This will be done in the database directly